src/util/security.rs

```diff
@@ -0,0 +1,202 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+use std::fs::File;
+use std::error::Error;
+use std::io::Read;
+use std::ptr;
+
+use grpc::{Channel, ChannelBuilder, ChannelCredentialsBuilder, ServerBuilder,
+           ServerCredentialsBuilder};
+
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+#[serde(default)]
+#[serde(rename_all = "kebab-case")]
+pub struct SecurityConfig {
+    pub ca_path: String,
+    pub cert_path: String,
+    pub key_path: String,
+    // Test purpose only.
+    #[serde(skip)]
+    pub override_ssl_target: String,
+}
+
+impl Default for SecurityConfig {
+    fn default() -> SecurityConfig {
+        SecurityConfig {
+            ca_path: String::new(),
+            cert_path: String::new(),
+            key_path: String::new(),
+            override_ssl_target: String::new(),
+        }
+    }
+}
+
+fn check_key_file(tag: &str, path: &str) -> Result<Option<File>, Box<Error>> {
+    if path.is_empty() {
+        return Ok(None);
+    }
+    match File::open(path) {
+        Err(e) => Err(
+            format!("failed to open {} to load {}: {:?}", path, tag, e).into(),
+        ),
+        Ok(f) => Ok(Some(f)),
+    }
+}
+
+fn load_key(tag: &str, path: &str) -> Result<Vec<u8>, Box<Error>> {
+    let mut key = vec![];
+    let f = check_key_file(tag, path)?;
+    match f {
+        None => return Ok(vec![]),
+        Some(mut f) => if let Err(e) = f.read_to_end(&mut key) {
+            return Err(
+                format!("failed to load {} from path {}: {:?}", tag, path, e).into(),
+            );
+        },
+    }
+    Ok(key)
+}
+
+impl SecurityConfig {
+    pub fn validate(&mut self) -> Result<(), Box<Error>> {
+        check_key_file("ca key", &self.ca_path)?;
+        check_key_file("cert key", &self.cert_path)?;
+        check_key_file("private key", &self.key_path)?;
+        // TODO: validate whether ca, cert and private key match.
+        if (!self.ca_path.is_empty() || !self.cert_path.is_empty() || !self.key_path.is_empty()) &&
+            (self.ca_path.is_empty() || self.cert_path.is_empty() || self.key_path.is_empty())
+        {
+            return Err("ca, cert and private key should be all configured.".into());
+        }
+
+        Ok(())
+    }
+}
+
+pub struct SecurityManager {
+    ca: Vec<u8>,
```

Comment:
why not using SecurityConfig directly? They have same field names.

Reply:
ping here

Reply:
They are not the same. `SecurityConfig` only stores the path, and `SecurityManager` stores the content. Separate them makes things clear and simple. Besides, it feels weird to call `bind` or `connect` method on a configuration struct. And `derive(Deserialize)` is conflict with `Drop` trait, which is the deal breaker.